packages/db/postgres/functions.sql

```diff
@@ -250,6 +250,95 @@ BEGIN
 END
 $$;
 
+DO 
+$$
+BEGIN
+  IF NOT EXISTS (SELECT 1 FROM pg_type WHERE typname = 'used_storage') THEN
+    CREATE TYPE used_storage AS (uploaded TEXT, pinned TEXT, total TEXT);
+  END IF;
+END
+$$;
+
+CREATE OR REPLACE FUNCTION user_used_storage(query_user_id BIGINT) 
+RETURNS used_storage
+LANGUAGE plpgsql
+AS
+$$
+DECLARE 
+  used_storage  used_storage;
+  uploaded      BIGINT;
+  pinned        BIGINT;
+  total         BIGINT;
+BEGIN
+  uploaded := 
+    (
+      SELECT COALESCE(SUM(c.dag_size), 0)
+      FROM upload u
+      JOIN content c ON c.cid = u.content_cid
+      WHERE u.user_id = query_user_id::BIGINT
+      AND u.deleted_at is null
+    );
+
+  pinned := 
+    (
+      SELECT COALESCE((
+        SELECT COALESCE(SUM(c.dag_size), 0)
+        FROM psa_pin_request psa_pr
+        JOIN content c ON c.cid = psa_pr.content_cid
+        JOIN pin p ON p.content_cid = psa_pr.content_cid
+        JOIN auth_key a ON a.id = psa_pr.auth_key_id
+        WHERE a.user_id = query_user_id::BIGINT
+        AND psa_pr.deleted_at is null
+        AND p.status = 'Pinned'
+        GROUP BY p.status
+      ), 0)
+    ); 
+
+  total := uploaded + pinned;
+
+  SELECT  uploaded::TEXT,
+          pinned::TEXT,
+          total::TEXT
+  INTO    used_storage;
+
+  return used_storage;
+END
+$$;
+
+CREATE OR REPLACE FUNCTION users_by_storage_used(
```

Comment:
My main concern here is performance :(

While this is ran as a cron job, I wonder if running this for a high number of users, and with users that  potentially have lots of files is going to be a problem.

I guess the former could be mitigated by sharding the operation, similarly to what we do in other cron jobs, not sure if there's an easy way to mitigate the latter?

I guess given we could monitor this and see the actual impact perfornance in prod?
Thoughts?

I wonder if makes us subject to DOs attacks?
ie. I create lots of really small files, I could be making this query be extremely expensive and have an impact to overal perf of DB?

Reply:
Agree, this is definitely something we should monitor in production to see if there are performance issues.